chore: adr for store enhancement (#933)
defenseunicorns/pepr: adr/0008-store-key-sanitization.md

# 8: Store Key Sanitization

Date: 2024-07-02

## Status

Under Review

## Context

> Issues were discovered when accepting arbitrary Keycloak client IDs and setting them into the Store without sanitization. For instance, IDs like `https://google.com/` failed during Store.setItem due to the `/` character, resulting in a `422` error.

Reviewer: ```suggestion
Issues were discovered when accepting arbitrary Keycloak client IDs as Store keys without sanitization. For instance, IDs like `https://google.com/` failed during `Store.setItem` due to the `/` character being interpreted as a path separator, resulting in a `422` error.
```

Similar to before, without clarifications like these there's no way for a future reader to really understand what's broken.

Author: yep good look, agree but respelled separator

## Decision

Implementing base64 would be a heavy shift for an extremely important part of the system. Instead, we will sanitize the key by replacing `/` with a character that does not break the `patch` operation before placing the key. We will also increase the tests to find other edge cases before releasing the change.

#### Sanitize using String Replacement

> * Sanitize the key by replacing `/` with a character that does not break the `patch` operation before placing the key.

Reviewer: Haven't read too deeply into the specs, but... since the patch operation is a [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902/#section-4), it's "data" key is defined as a [JSON Pointer](https://datatracker.ietf.org/doc/html/rfc6901#section-3) object... which means that both "~" and "/" are going to have special meanings:

```
A JSON Pointer is a Unicode string (see [[RFC4627], Section 3](https://datatracker.ietf.org/doc/html/rfc4627#section-3))
   containing a sequence of zero or more reference tokens, each prefixed
   by a '/' (%x2F) character.

   Because the characters '~' (%x7E) and '/' (%x2F) have special
   meanings in JSON Pointer, '~' needs to be encoded as '~0' and '/'
   needs to be encoded as '~1' when these characters appear in a
   reference token.
   ```

Author: added this link and spoke to the replacement but kept specific implementation as TBD by the testing done during the implementation.

> Sanitize the keys by replacing `/` with [a character that does not break the `patch` operation](https://datatracker.ietf.org/doc/html/rfc6902/#section-4) before setting, getting, or deleting a key in/from the store. We will also increase the tests to find other edge cases before releasing the change and deciding upon the exact character or pattern. We are prioritizing the cluster operator persona in this case who wants to quickly check all keys and values in the store by looking at the PeprStore CR.

Reviewer: ```suggestion
Sanitize the keys by replacing `/` with [one or more characters](https://datatracker.ietf.org/doc/html/rfc6901#section-3) that do not break the [`patch` operation](https://datatracker.ietf.org/doc/html/rfc6902/#section-4) before setting, getting, or deleting a key in/from the store. We will also increase the tests to find other edge cases before releasing the change and deciding upon the exact character or pattern. We are prioritizing the cluster operator persona in this case who wants to quickly check all keys and values in the store by looking at the PeprStore CR.
```

> * Increasing the tests to find other edge cases before releasing the change.

Reviewer: It might be nice to think through / add some examples of the sorts of keys we'd like the Store to be able to handle in order to give "other edge cases" (and their respective, future tests) some shape.  Like:
- URIs (cuz that's what drove this PR to begin with)
- dot-delimited strings -- e.g. "a.b.c.d" -- just to be sure our escaping can handle things that look like JSON
- " " (empty strings)
- "null" (JS keywords)
- ...and so on.

Author: I hear you on URIs but Store.*Item only accepts strings. 

It can already handle dot-delimited and empty strings. 

The only error we know of until doing more testing and finding more would be url as a string.

```yaml
  data:
    __pepr_do_not_delete__: k-thx-bye
    hello-pepr-: b #the key is empty here
    hello-pepr-a.a.a: a
  kind: PeprStore
```

```ts
OnSchedule({
  name: "set-items",
  unit: "seconds",
  every: 15,
  run: () => {
    Store.setItem("a.a.a","a")
    Store.setItem("","b")
  }
})
```
